source/common/ssl/connection_impl.cc

```diff
@@ -251,8 +251,21 @@ std::string ConnectionImpl::subjectPeerCertificate() {
   if (!cert) {
     return "";
   }
-  bssl::UniquePtr<char> buf(X509_NAME_oneline(X509_get_subject_name(cert.get()), nullptr, 0));
-  return std::string(buf.get());
+
+  // Format settings for a single line in RFC 2253 format.
+  constexpr int indent = 0;
+  constexpr unsigned long flags = XN_FLAG_RFC2253;
+
+  bssl::UniquePtr<BIO> buf(BIO_new(BIO_s_mem()));
+  X509_NAME_print_ex(buf.get(), X509_get_subject_name(cert.get()), indent, flags);
+  const uint8_t* buf_data;
+  size_t buf_data_len;
+  int mem_content_rc = BIO_mem_contents(buf.get(), &buf_data, &buf_data_len);
```

Comment:
`BIO_mem_contents` doesn't allocate or copy. It copies the internal data pointer for the BIO into its parameters, semantically equivalent to `buf_data = ((BUF_MEM *)bio->ptr)->data` .